[WIP] Add ui:clean task

NickLaMuro opened the pull request:
Cleans up `node_modules` and `vendor/assets` directories of ALL `manageiq-ui-classic` gem directories in your gem dir.  Takes a few small bits of `Bunlder::Runtime.clean` to for some inspiration, but has a slightly different take on it (similar to `brew cleanup`).


Overview
--------

This task takes a `Dir.glob` of the git gems that match `manageiq-ui-classic`, minus the current one (as determined by using the `Bundler.definition.spec_git_paths` and selecting `manageiq-ui-classic`), and looping through each, removing any `node_modules/` and `vendor/assets/` directories that exist in each.  This keeps the gems in place, but just removes any compiled JS files when running a `rake update:ui` when the gem is active.

As each dir is purged, the output of savings will be noted prior to purging the directory, and a total savings displayed at the end.


Usage
-----

This task can be run in two modes:

* Normal:
  
  ```console
  $ bundle exec rake ui:clean
  ```
  
* Dry Run:  Does everything minus the actual deletes
  
  ```console
  $ DRY_RUN=1 bundle exec rake ui:clean
  ```
  
The code would be much smaller if I didn't include the disk usage bits, but I wanted to be fancy.

Sample output:

```console
$ bundle exec rake ui:clean
Removing bundler/gems/manageiq-ui-classic-e36308c72b24/node_modules... (193.4MB)
Removing bundler/gems/manageiq-ui-classic-e36308c72b24/vendor/assets... (123.8MB)
Removing bundler/gems/manageiq-ui-classic-e7f647e32eb9/vendor/assets... (0.2MB)
Removing bundler/gems/manageiq-ui-classic-ed1f7171072a/vendor/assets... (0.2MB)
Removing bundler/gems/manageiq-ui-classic-ee408923680d/vendor/assets... (0.2MB)
Removing bundler/gems/manageiq-ui-classic-eedbe186e2b6/vendor/assets... (0.2MB)
Removing bundler/gems/manageiq-ui-classic-f4b6884536c1/vendor/assets... (76.2MB)
Removing bundler/gems/manageiq-ui-classic-f687d09fa2ad/vendor/assets... (0.2MB)

Total Savings: 394.4MB
```


Links
-----

* Issue brought up in chat:  https://gitter.im/ManageIQ/manageiq/ui?at=5aaaa9b40a1614b71223a661


Testing for yourself
-------------------

I was testing this myself by overriding the gem, bundling, and running `rake ui:clean`, but should be able to be run from the `manageiq-ui-classic` directory as well, it just won't filter out the `manageiq-ui-classic` in the bundler definition (now that I think about it... not sure how I was getting a skipped dir in this setup... oh well).

NickLaMuro commented:
cc @djberg96 @chrisarcand 

To the reviewer that gets this:  Check out the link in the description to where I explain the issue in gitter.  Feel free to chuck this if you don't want this code in the project, I just thought I would share since it was a convenient way to clean things up and see the total savings.

djberg96 commented:
@NickLaMuro Nice work!

NickLaMuro commented:
Update:  Fixed some of the rubocop issues after this fix, but didn't get a chance to test that I didn't break anything in doing so.  Will try to do that at some point soon.

chrisarcand commented:
I mean this is great for the purpose you wrote it for but what's your reasoning for not just using `bundle clean` (or our own purposed command here simply utilizing it)? Is it nice that you're keeping the gems and just removing the compiled JS? Why?

NickLaMuro commented:
@chrisarcand Totally crossed my mind as well, and am on the fence on whether or not doing that makes more sense or not.

My biggest reason for going this route is for audit reasons.  If a dev ran into an issue with a certain issue of this gem (on the ruby side), ran a `bundle update`, and then the issue fixed itself, we would be able to see find the offending version ala `ls -lh /path/to/bundle/gem | grep manageiq-ui-classic`.  If a developer runs in to the same issue you had, you can help determine the culprit version of the gem.  Doing a `bundle clean` obliterates that as a possibility.

In some respects, this is just doing a `gem pristine`, on all of the non-used gems, but I was being pedantic in not wanting to spawn subprocesses probably do what was effectively a `system("git clean")` for every version of the gem... because #Reasons™.  That said, not sure the complexity is worth that niche case, as you have already determined.

* * *

Long story short, I mostly opened this because it was useful to me when I realized there was a non-trival amount of my disk being used for node packages I was not actively using.  The codebase can definitely survive without it, and I can make this work for just be as a custom `config/initializer/my_super_awesome_stuff.rb` in my `manageiq` repo, so I won't have any hard feelings if we decide to not include this.

(okay, I lied, I will cry myself to sleep at night over this if you don't merge😭 )

chrisarcand commented:
> In some respects, this is just doing a gem pristine, on all of the non-used gems, but I was being pedantic in not wanting to spawn subprocesses probably do what was effectively a system("git clean") for every version of the gem... because #Reasons™. 

Huh? Elaborate more? That method actually sounds really nice to me, running pristine on all MIQ gems sounds Good®

NickLaMuro commented:
@chrisarcand I might be confused on your confusion, but I will elaborate anyway:

Under the covers, I am pretty sure that `bundler` will know that gems in the `bunlder/gems` dir will be `git` gems (too lazy to look it up), so doing a `gem pristine` wouldn't work in the traditional sense.  But I think it has a work around for it (pretty sure there is a `bundler pristine`... yup... [looked it up](http://bundler.io/v1.15/man/bundle-pristine.1.html) ).

I don't like calling subprocesses in ruby when I can avoid it (again, reasons a pretty pedantic because I don't like subprocesses and the cost associated with them), and since there were two known file directories, I opted to use `FileUtils#rm_rf` instead.  That said, it would have relatively the same effect with in reality no-difference in execution time, just with less "Nick's Patented Pedantic Sauce" spread all over it.

Does that make sense?

chrisarcand commented:
Uh sure! I'll leave it to others to decide a path moving forward (because...Reasons™ :) ) but IMO if it really is just your pedantic sauce not wanting to use subprocesses I'd prolly go that route instead. 

Though, is there any API in Bundler to utilize the functionality directly without shelling out a process and thus addressing all concerns?

NickLaMuro commented:
> Though, is there any API in Bundler to utilize the functionality directly without shelling out a process and thus addressing all concerns?

This I know from experience, but everything `git` related in bundler assumes a `git` binary on the OS `$PATH` (or `%PATH%` on Windows).  I... `git`... why they did that, since they tried to limit the number of dependencies shipped with `bundler` (by that, I mean the ones they do have are all vendored in via `automatiek` rake tasks), and vendoring `rugged` would mean shipping with a `c-ext` I am pretty sure... so that probably is a no go.

Fryguy commented on lib/tasks/manageiq/ui_tasks.rake:
```diff
@@ -24,6 +24,100 @@ namespace :update do
+        path = Pathname.new(file_path_to_delete)
+        next unless path.exist?
+
+        # Disk usage code ulled from Homebrew.  License for Homebrew below:
```
Typo `ulled`?  is it `pulled`?

NickLaMuro replied:
Yup, typo.

Fryguy commented on lib/tasks/manageiq/ui_tasks.rake:
```diff
@@ -24,6 +24,100 @@ namespace :update do
+              "#{'%.1f' % (total_savings.to_f / 1_048_576)}MB"
+            end
+
+    puts "\nTotal Savings: #{total}"
```
Do we care about this?  I wonder if it's worth having it considering that we have to pull over the license.

NickLaMuro replied:
> Do we care about this?

The savings portion of the code or the entire PR in general?

:trollface: (sort of)

To your actual question, to some degree I think it shines light on to a major issue (to me at least) of bloat that can happen in the git gems and possibly wanting to re-structure where we put these assets so they aren't ending up being dumped into the git gem dirs.

That said, does it actually help with the effectiveness of the script?  No.  Also, it probably makes it slower, but I thought it was neat and agree that it probably isn't worth dragging in a license for that reason alone.

* * *

To the other question, can we live without this?  Probably.  I had a non-trival amount of space sucked up by `node-modules` dirs (about 5 gigs) prior to making this script, and figured I could write something that people might find useful and determine if it was for more folks than myself.

As discussed with Chris, there are other ways that this can be accomplished, some of which don't even require this script (`bundle clean`, for example) and so not merging this for that reason alone seems pretty reasonable.  Also, we can do this in a way that doesn't require polluting the production build by having an ancillary gem added to peoples `bundle` via `bundler.d` that injects this task as well, so I would be cool with going that route as well.

himdel commented on lib/tasks/manageiq/ui_tasks.rake:
```diff
@@ -24,6 +24,100 @@ namespace :update do
+    stale_ui_gems.each do |stale_miq_ui_gem_dir|
+      [
+        File.join(stale_miq_ui_gem_dir, "node_modules"),
+        File.join(stale_miq_ui_gem_dir, "vendor", "assets")
```
can this be +`, "bower"`?

We did have some vendor files which were not bower at some point..

NickLaMuro replied:
@himdel Addressed this in my rebase, and I think the `gem_regexp` should do the trick... but the sad part was:

```
DRY RUN:  NO OPERATIONS BEING EXECUTED!!
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-graphql-4b9ec2244e05/node_modules... (15.6MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-graphql-55a4ec981f49/node_modules... (20.6MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-graphql-d08c7ff58785/node_modules... (14.2MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-265ba60b2652/node_modules... (193.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-265ba60b2652/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-61026e3296a1/node_modules... (195.3MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-61026e3296a1/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-6805dd8a1aa7/node_modules... (143.7MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a0ff186c3fef/node_modules... (186.2MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a0ff186c3fef/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a63c3559bb12/node_modules... (186.2MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/manageiq-ui-classic-a63c3559bb12/vendor/assets/bower... (123.8MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/miq_v2v_ui_plugin-761f3ad07ef3/node_modules... (206.5MB)
Removing ~/.gem/ruby/2.3.3/bundler/gems/miq_v2v_ui_plugin-cd901ff708d5/node_modules... (204.1MB)

Total Savings: 1.8GB
```

And I cleaned this up like 3 weeks ago... 😭

himdel replied:
Ouch :)

The fun thing is, it only happens when using the gems, so I mostly blame bundler for putting each version in a new directory and not cleaning up the rest. (That said, it's definitely wrong to install `node_modules` under a gem dir too.)

My personal solution to this has been:

`30 14 * * * hardlink -pot ~/.rbenv/ ~/ManageIQ/`

in my `crontab`, but looks like we should really look into moving all the `node_modules` under `manageiq`.

(My guess about the dependency ordering here will be: finish the move to webpack5, try to do that, realize npm can't actually install to any directory other than <where package.json is>/node_modules and while require/import have the same problem this can be overriden in weback loader configs .. which will likely only work for JS (babel), not typescript, and then spend some time figuring out whether typescript loaders can do the same or whether we need to pass ts through babel as well and just use tsc as a linter :).)

himdel commented:
@NickLaMuro for symmetry, I think this should not only clean ui-classic, but also all the UI plugins.

I mean, `update:ui` does run `yarn` in all the UI plugin repos, so presumably `ui:clean` should delete it?

himdel commented:
@miq-bot add_label wip

NickLaMuro commented:
Hey @himdel , sorry for not answering your question from above:

> I mean, update:ui does run yarn in all the UI plugin repos, so presumably ui:clean should delete it?

Makes sense to me.  Do you see value you in this PR being revisited and should I get this working?  I haven't touched it since I was unsure if this was something people would find value out of, but if it is something you would like to see merged in, I can make it happen.

NickLaMuro commented:
Going to close this for now.  I assume if this is something people wanted, those following this PR can speak up.
